File: backend/spec/requests/api/v1/stages_spec.rb
require "rails_helper"

RSpec.describe "Api::V1::Stages" do
  before do
    Stage.create!(slug: "database/repl", part: "database", stage_number: 1, title: "User REPL", description: "Build a REPL.")
    Stage.create!(slug: "database/lexer", part: "database", stage_number: 2, title: "SQL Lexer", description: "Tokenize SQL.")
  end

  describe "GET /api/v1/stages" do
    it "returns all stages ordered by stage_number" do
      get "/api/v1/stages"

      expect(response).to have_http_status(:ok)
      json = response.parsed_body
      expect(json.length).to eq(2)
      expect(json[0]["slug"]).to eq("database/repl")
      expect(json[1]["slug"]).to eq("database/lexer")
    end

    it "does not require authentication" do
      get "/api/v1/stages"
      expect(response).to have_http_status(:ok)
    end
  end

  describe "GET /api/v1/stages/:part/:stage" do
    it "returns a specific stage" do
      get "/api/v1/stages/database/repl"

      expect(response).to have_http_status(:ok)
      json = response.parsed_body
      expect(json["slug"]).to eq("database/repl")
      expect(json["title"]).to eq("User REPL")
    end
  end
end

File: e2e/tests/stages.spec.js
import { test, expect } from "@playwright/test";

const API_BASE = process.env.API_URL || "http://localhost:4000";

test.describe("Stages API", () => {
  test("GET /api/v1/stages returns all stages", async ({ request }) => {
    const response = await request.get(`${API_BASE}/api/v1/stages`);

    expect(response.status()).toBe(200);
    const stages = await response.json();
    expect(stages.length).toBeGreaterThanOrEqual(12);

    const repl = stages.find((s) => s.slug === "database/repl");
    expect(repl).toBeTruthy();
    expect(repl.title).toBe("User REPL");
    expect(repl.stage_number).toBe(1);
  });

  test("GET /api/v1/stages/database/repl returns stage details", async ({ request }) => {
    const response = await request.get(`${API_BASE}/api/v1/stages/database/repl`);

    expect(response.status()).toBe(200);
    const stage = await response.json();
    expect(stage.slug).toBe("database/repl");
    expect(stage.part).toBe("database");
  });

  test("GET /health returns ok", async ({ request }) => {
    const response = await request.get(`${API_BASE}/health`);
    expect(response.status()).toBe(200);
    const body = await response.json();
    expect(body.status).toBe("ok");
  });
});
